### ISSUE TITLE
Principal wallet migration strategy, take 2

### ISSUE THREAD

**Issue by rbrunner7** [seraphis-migration/wallet3#19]
@DangerousFreedom1984 already touched the question of the principal wallet migration strategy to apply in his issue #10, but not much about that was discussed there as the issue mutated into a discussion of serialization. That's why I come back to the subject with a new, separate issue.

In the wiki you find a text of mine from February about this: [Principal Wallet Migration Strategies](https://github.com/seraphis-migration/strategy/wiki/Principal-Wallet-Migration-Strategies).

As I see it, what I wrote there still more or less stands: There are only 3 principal wallet migration strategies that make sense. Thus for further discussion I assume knowledge about that text.

I have updated the graphic however to reflect the situation as it turned out, with @UkoeHB writing a very comprehensive library that implements Seraphis and Jamtis, but not a wallet proper as originally planned.

![Wallet3_updated](https://user-images.githubusercontent.com/10373537/196025709-091c89e1-f9e7-4133-82a4-113ffb99a578.png)

**Comment by rbrunner7** [comment]
My comments about the *wallet2+* strategy which would basically mean to add Seraphis and Jamtis support to the existing codebase, i.e. `wallet2.h` and `wallet2.cpp`, with the least possible amount of disruption and with letting the current API mostly stand as it is today:

It is common knowledge among Monero devs that `wallet2.cpp` is too big and too complicated by far for comfort, and without any future technology on the horizon like Seraphis we probably should start a project for comprehensive refactoring and modularization of that code right away.

I had to work with that code more closely myself when I implemented [this Monero PR](https://github.com/monero-project/monero/pull/8076) which adds a new feature which is not too big and not terribly complicated but which already was quite difficult to build exactly because the code in `wallet2.cpp` is such a tangle.

For me adding Seraphis and Jamtis support to the monolithic `monero2` of today is out of question: Nobody wants those files even bigger and even more complicated.

So how about refactoring first, and then add Seraphis and Jamtis? That could technically work, but I really hesitate to "pump" so much work into what is quite old technology. Refactoring first and only then go on and start adding those could also take very long, maybe substantially longer than building something new.

Furthermore, as an API, the design of `wallet2.h` is not very good IMHO, which is of course no surprise if you see how much larger it became over the years, and how many different people worked on that. I think we can do much better if we get the freedom to define a new API from scratch.

Thus for me, without clear (and for me surprising) votes of most other project members or important wallet app coders in favor of the *wallet2+* strategy, this is already more or less dead.

**Comment by rbrunner7** [comment]
My comments about the *wallet3* strategy:

When I came up with the idea of a "unifying" wallet API half a year ago speed was an important element: Make a first version of *wallet3* available early, so wallet app writers can start rewriting their wallets early. The resulting time-wise parallelization of work - rewriting apps and developing the Seraphis wallet could run in parallel - would hopefully make possible an earlier hardfork to Seraphis.

Trouble is: Right now I have serious doubts whether it will be possible for me, or anybody else for that matter, to define `wallet3.h` anytime soon. The main reason is that we don't know yet many important aspects of the Seraphis wallet: How it will look, how it will work, how its "natural" interface will look, how the "jobs" will be distributed between the Seraphis library and the Seraphis wallet, etc.

And as long as the Seraphis wallet is "in the dark" to this degree, it seems hard to define a unifying interface that will be able to cover and support it. And another question also has to wait for an answer: Are `wallet2` and the Seraphis wallet similar enough that such a unification is feasible with reasonable effort in the first place?

IMO this means that *wallet3* as a strategy is still a possibility, but it's not possible to commit to it now or in the near future.

What we can do is keeping an eye on it as things about the Seraphis wallet will progressively clear and take shape over the coming months.

**Comment by DangerousFreedom1984** [comment]
I guess that in the end we will need to have something similar as 'wallet3' like you propose. For now, I think we should focus on working on the basic functions in order to make seraphis usable. So both approaches 'double wallet' and 'wallet3' seems reasonable to me as long as we keep good coding practices like modularity.
